feat(web): add support for casting

bwees opened the pull request:
## Description

This adds support for Google Cast and [FCast](https://fcast.org). This works with the photo viewer, video viewer, and also works with slideshow mode.

## How Has This Been Tested?

This has been tested with a Google TV Streamer and with Chrome on MacOS and Windows.


## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

bwees committed:
recreate #13966

bwees committed:
gcast button works

bwees committed:
rewrote gcast-player to be GCastDestination and CastManager manages the interface between UI and casting destinations

bwees committed:
remove unneeded imports

bwees committed:
add "Connected to" translation

bwees committed:
Remove css for cast launcher

bwees committed:
fix tests

bwees committed:
fix doc tests

bwees committed:
fix the receiver application ID

bwees committed:
remove casting app ID

alextran1502 commented:
Looks good! Instead of having the cast button on the navigation bar on every page. I think it makes sense to have it in the 

- Asset viewer
- Album view
- Shared link view

With all other views, the users can start casting from the asset viewer

bwees committed:
remove cast button from nav bar

It is now present at the following locations:

- shared link album and single asset views
- asset viewer (normal user)
- album view (normal user)

danieldietzler commented on web/package.json:
```diff
@@ -94,6 +95,7 @@
     "typescript": "^5.7.3",
     "typescript-eslint": "^8.28.0",
     "vite": "^6.0.0",
+    "vite-tsconfig-paths": "^5.1.4",
```
This isn't used, is it?

danieldietzler commented on web/src/lib/cast/cast-button.svelte:
```diff
@@ -0,0 +1,31 @@
+  <CircleIconButton
+    color={getButtonColor()}
+    icon={castManager.isCasting ? mdiCastConnected : mdiCast}
+    onclick={() => void GCastDestination.showCastDialog()}
```
```suggestion
    onclick={GCastDestination.showCastDialog}
```
I think the button should be fine with this...?

danieldietzler commented on web/src/lib/cast/cast-button.svelte:
```diff
@@ -0,0 +1,31 @@
+  };
+</script>
+
+{#if castManager.availableDestinations.length > 0 && castManager.availableDestinations[0].type == CastDestinationType.GCAST}
```
```suggestion
{#if castManager.availableDestinations.length > 0 && castManager.availableDestinations[0].type === CastDestinationType.GCAST}
```

danieldietzler commented on web/src/lib/components/album-page/album-viewer.svelte:
```diff
@@ -103,6 +104,8 @@
       {/snippet}
 
       {#snippet trailing()}
+        <CastButton isWhiteHover={true} />
```
```suggestion
        <CastButton isWhiteHover />
```

danieldietzler commented on web/src/lib/components/asset-viewer/photo-viewer.svelte:
```diff
@@ -145,6 +146,20 @@
 
+  $effect(() => {
+    if (assetFileUrl) {
+      void cast(assetFileUrl);
```
Like this you basically lose any chance to provide the user with an helpful error. We have a `handleError` (or something like that) function in web that handles promises. I think that would be preferred here.

danieldietzler commented on web/src/lib/components/asset-viewer/video-remote-viewer.svelte:
```diff
@@ -0,0 +1,95 @@
+  let previousPlayerState: CastState | null = $state(null);
+
+  const handlePlayPauseButton = () => {
+    if (castManager.castState === CastState.PLAYING) {
+      castManager.pause();
+    } else if (castManager.castState === CastState.IDLE) {
+      void cast(assetFileUrl, true);
+    } else {
+      castManager.play();
+    }
```
More of a personal preference, but a switch could be nice here too. Especially if there may be more states eventually (idk maybe something like PAUSE exists in other case implementations?)

danieldietzler commented on web/src/lib/components/asset-viewer/video-remote-viewer.svelte:
```diff
@@ -0,0 +1,95 @@
+    if (castManager.castState === CastState.PLAYING) {
+      castManager.pause();
+    } else if (castManager.castState === CastState.IDLE) {
+      void cast(assetFileUrl, true);
```
Try to not just void every promise by default. The default should be to await a promise, unless there's a good reason not to.

danieldietzler commented on web/src/lib/components/asset-viewer/video-remote-viewer.svelte:
```diff
@@ -0,0 +1,95 @@
+
+  $effect(() => {
+    if (
+      castManager.castState !== previousPlayerState &&
```
This case is impossible, right? This effect will only trigger *if* `castState` (or `previousPlayerState` I guess) has changed, in which case it isn't the previous state anymore.

danieldietzler commented on web/src/lib/components/asset-viewer/video-remote-viewer.svelte:
```diff
@@ -0,0 +1,95 @@
+      return;
+    }
+    const fullUrl = new URL(url, globalThis.location.href);
+    const didCast = await castManager.loadMedia(fullUrl.href, force);
```
What does this variable specifically mean? Does it resolve when the cast has been aborted?

danieldietzler commented on web/src/lib/components/asset-viewer/video-remote-viewer.svelte:
```diff
@@ -0,0 +1,95 @@
+  };
+
+  function handleSeek(event: Event) {
+    const newTime: number = Number.parseFloat((event.target as HTMLInputElement).value);
```
```suggestion
    const newTime = Number.parseFloat((event.target as HTMLInputElement).value);
```
I would expect the typing of `parseFloat` to be correct?

danieldietzler commented on web/src/lib/managers/cast-manager.svelte.ts:
```diff
@@ -0,0 +1,163 @@
+  availableDestinations = $state<CastDestination[]>([]);
+  initialized = $state(false);
+
+  isCasting = $derived<boolean>(this.current !== null && this.current.isConnected);
```
`this.current?.isConnected`?

danieldietzler commented on web/src/lib/utils/cast/gcast-destination.svelte.ts:
```diff
@@ -0,0 +1,264 @@
+  type = CastDestinationType.GCAST;
+  isAvailable = $state(false);
+  isConnected = $state(false);
+  currentTime: number | null = $state(null);
+  duration: number | null = $state(null);
+  castState = $state(CastState.IDLE);
+  receiverName: string | null = $state(null);
```
`$state` also accepts a generic, which (IMO) is cleaner, so like `currentTime = $state<number | null>(null);`

danieldietzler commented on web/src/lib/utils/cast/gcast-destination.svelte.ts:
```diff
@@ -0,0 +1,264 @@
+
+    console.debug('GCast API available:', this.isAvailable);
+
+    if (this.isAvailable) {
```
Maybe early return here with `if (!this.available) { return false; }`?

danieldietzler commented on web/src/lib/utils/cast/gcast-destination.svelte.ts:
```diff
@@ -0,0 +1,264 @@
+        autoJoinPolicy: chrome.cast.AutoJoinPolicy.ORIGIN_SCOPED,
+      });
+
+      //
+      // setup the event listeners
+      //
```
Such comments are not really necessary. I think comments generally make sense if, for instance, you made some assumptions that aren't super explicit, or if a code section is really ugly because x requires this. Just explaining the code below doesn't really add anything (neither does commenting complex code. if it's so complex that it needs a comment, you should really refactor it).

danieldietzler commented on web/src/lib/utils/cast/gcast-destination.svelte.ts:
```diff
@@ -0,0 +1,264 @@
+    let contentType: string | null = null;
+
+    // in the future we can swap this out for an API call to get image metadata
+    await fetch(mediaUrl, { method: 'HEAD' }).then((response) => {
+      contentType = response.headers.get('content-type');
+    });
```
Maybe I am too tired right now, but what are you awaiting here?

bwees replied:
This is to grab the content-type for the thumbnail. @jrasm91 would it be too hard to get the thumbnail contentType with the asset info, would come in handy on mobile to keep it using openapi calls.

danieldietzler reviewed (commented):
The code looks already pretty great actually! Really just nitpicks for the most part :)

bwees committed:
part 1 of fixes from @danieldietzler code review

jrasm91 commented on web/src/lib/utils/cast/gcast-destination.svelte.ts:
```diff
@@ -0,0 +1,264 @@
+import 'chromecast-caf-sender';
+import { Duration } from 'luxon';
+
+const FRAMEWORK_LINK = 'https://www.gstatic.com/cv/js/sender/v1/cast_sender.js?loadCastFramework=1';
```
Is it possible to bundle this file in our build instead of having the application call out to google whenever the cast button is used?

bwees replied:
We could but the recommendation from google is that you load it directly. Do you want me to add it?

bwees replied:
@alextran1502 agreed that we should just keep this as is since the cast framework calls out to google anyway regardless if we host this file ourselves.

bwees commented on web/src/lib/cast/cast-button.svelte:
```diff
@@ -0,0 +1,31 @@
+</script>
+
+{#if castManager.availableDestinations.length > 0 && castManager.availableDestinations[0].type == CastDestinationType.GCAST}
+  <CircleIconButton
```
It does not match the other buttons in the tab bar when hovering if I do this. Once the tab bar buttons are migrated I will do this (or it will happen in migration)

bwees committed:
part 2 of code review changes from @danieldietzler and @jsram91

bwees committed:
cleanup documentation

bwees committed:
onVideoStarted missing callback

bwees committed:
add token expiry validation

bwees committed:
cleanup logic and logging

bwees committed:
small cleanup

alextran1502 reviewed (approved):
Thank you, working well functionally! One nit

bwees committed:
rename to ICastDestination

bwees committed:
cast button changes

jrasm91 commented on web/src/lib/cast/cast-button.svelte:
```diff
@@ -0,0 +1,45 @@
+      aria-label={$t('cast')}
+    />
+  {:else}
+    <CircleIconButton
```
Why is this the old button still?

bwees replied:
This was explained here: https://github.com/immich-app/immich/pull/18231#discussion_r2096504982

jrasm91 replied:
Sounds good! Sorry I missed that reply on mobile 😅

jrasm91 commented on web/src/lib/managers/cast-manager.svelte.ts:
```diff
@@ -0,0 +1,159 @@
+  private sessionKey: SessionCreateResponseDto | null = null;
+
+  constructor() {
+    // load each cast destination
```
Why are all these comments still here?

bwees replied:
I felt that they were needed to clarify behavior. I deleted most of the others that were just markers for sections of code. In the future I will try and be more succinct/sparing in my commenting

jrasm91 replied:
Personally I think almost every one of these comments can be removed and don't add any value to the code. Many just duplicate the meaning implied from the variable and method names
